tests/test_templatetags_slot_fill.py

```diff
@@ -481,98 +698,297 @@ def test_component_without_default_slot_refuses_implicit_fill(self):
+
+                {% if 1 > 2 %}
+                    {% fill "footer" %}
+                        FOOTER
+                    {% endfill %}
+                {% endif %}
+            {% endcomponent %}
+        """
+        template = Template(template_str)
+
+        with self.assertRaisesMessage(TemplateSyntaxError, "Illegal content passed to component 'test'"):
+            template.render(Context({"slot_names": ["header", "main"]}))
+
+    @parametrize_context_behavior(["django", "isolated"])
+    def test_slots_inside_loops(self):
+        @register("test_comp")
+        class OuterComp(Component):
+            def get_context_data(self, name: Optional[str] = None) -> Dict[str, Any]:
+                return {
+                    "slots": ["header", "main", "footer"],
+                }
+
+            template: types.django_html = """
+                {% load component_tags %}
+                {% for slot_name in slots %}
+                    <div>
+                        {% slot name=slot_name %}
+                            {{ slot_name }}
+                        {% endslot %}
+                    </div>
+                {% endfor %}
+            """
 
         template_str: types.django_html = """
             {% load component_tags %}
-            <body>
-                {% component "test" %}
-                    ABC: {{ name }} {{ some }}
-                {% endcomponent %}
-            </body>
+            {% component "test_comp" %}
+                {% fill "header" %}
+                    CUSTOM HEADER
+                {% endfill %}
+                {% fill "main" %}
+                    CUSTOM MAIN
+                {% endfill %}
+            {% endcomponent %}
         """
-        self.template = Template(template_str)
+        template = Template(template_str)
+        rendered = template.render(Context())
 
-        nested_ctx = Context()
-        # Check that the component can access vars across different context layers
-        nested_ctx.push({"some": "var"})
-        nested_ctx.push({"name": "carl"})
-        rendered = self.template.render(nested_ctx)
+        expected = """
+            <div>CUSTOM HEADER</div>
+            <div>CUSTOM MAIN</div>
+            <div>footer</div>
+        """
+        self.assertHTMLEqual(rendered, expected)
 
-        self.assertHTMLEqual(
-            rendered,
+    @parametrize_context_behavior(["django", "isolated"])
+    def test_passthrough_slots(self):
+        registry.register("slotted", SlottedComponent)
+
+        @register("test_comp")
+        class OuterComp(Component):
+            def get_context_data(self, name: Optional[str] = None) -> Dict[str, Any]:
+                return {
+                    "slots": self.input.slots,
+                }
+
+            template: types.django_html = """
+                {% load component_tags %}
+                <div>
+                    {% component "slotted" %}
+                        {% for slot_name in slots %}
```

Comment:
One nice side-effect of this new render-time slot and fill resolution is that we can now allow ANY tags to be inside the `{% component %} {% endcomponent %}` tags, **as long as they don't output any text.**

So users can use for-loops to generate the fills:

```django
{% component %}
  {% for item in items %}
    {% fill name=item %}
      {{ item }}
    {% endfill %}
  {% endfor %}
{% endcomponent %}
```

But also e.g. `{% with %}`:

```django
{% component %}
  {% with abc="xyz" %}
    {% fill name=abc %}
      {{ abc }}
    {% endfill %}
  {% endwith %}
{% endcomponent %}
```

And even their custom tags, or using the `as var` syntax:

```django
{% component %}
  {% expr "['header', 'main', 'footer']" as slots %}
  {% for slot in slots %}
    {% fill name=slot %}
      {{ slot }}
    {% endfill %}
  {% endfor %}
{% endcomponent %}
```